### PR TITLE
vector based string implementation

### PR DESCRIPTION
**Pull request by arccth**
clean pull request this time

### CODE REVIEW

File: course4/homework-string/string.cc [hunk 1]
```diff
@@ -6,4 +6,76 @@ string::string()
 {
 }
 
-}
\ No newline at end of file
+string::string(const char * c)
+{
+    int i = 0;
+    while(c[i] != '\0') // I guess we are not supposed to use strlen
```
Comment (maciekgajewski) [inline course4/homework-string/string.cc:12]: You are _expected_ to use `std::strlen`

File: course4/homework-string/string.cc [hunk 2]
```diff
@@ -6,4 +6,76 @@ string::string()
...
+}
+
+string::string(const string& source)
+{
+    chars = source.chars;
+}
+
+string::string(string&& source)
+{
+    chars = std::move(source.chars);
```
Comment (maciekgajewski) [inline course4/homework-string/string.cc:26]: Is there a point implementing these constructors? The default implementations would do exactly the same.

File: course4/homework-string/string.hh [hunk 3]
```diff
@@ -1,10 +1,47 @@
...
+
+    char& operator [](const int index);
+
+    string& operator =(const string& source);
+
+    string& operator=(string&& source);
+
+    friend std::ostream& operator<<(std::ostream& os, const string& s)
+    {
+        for(int i=0; i < s.chars.size(); ++i)
```
Comment (maciekgajewski) [inline course4/homework-string/string.hh:39]: Use range-based for

**Review by maciekgajewski — COMMENTED** [review]
It is, indeed, too easy.
And your implementation would be slower than a simple, unique_ptr-based one. Why?

**Comment by arccth** [comment]
Created unique_ptr implementation here: https://github.com/maciekgajewski/c-excercises/pull/61
I guess it will be slower because multiple push to vector is more expensive than allocation of known size.

**Pull request closed without merging**